gapic-generator/lib/gapic/resource_lookup.rb

```diff
@@ -0,0 +1,83 @@
+# frozen_string_literal: true
+
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+module Gapic
+  module ResourceLookup
+    def self.for_service service
+      file = service.parent
+      api = file.parent
+
+      resource_types = []
+      service.methods.map do |method|
+        resource_types += get_message_resource_types method.input
+      end
```

Comment:
Instead of the above, maybe:
```ruby
resource_types = service.methods.flat_map do |method|
  get_message_resource_types method.input
end
```

(Or at least change `map` to `each` if you want to concatenate explicitly.)

Indeed, we can probably remove the need for a lot of the `flatten` calls around here, by using `Enumerable#flat_map` where appropriate.

Reply:
Refactored.